Sources/PointFree/Episode/Show.swift

```diff
@@ -38,7 +38,125 @@ let episodeResponse: Middleware<StatusLineOpen, ResponseEnded, Tuple4<Either<Str
     }
 )
 
-let episodeView = View<(Database.User?, Stripe.Subscription.Status?, Episode)> { user, subscriptionStatus, episode in
+let useCreditResponse =
+  filterMap(
+    over1(episode(forParam:)) >>> require1 >>> pure,
+    or: writeStatus(.notFound) >-> respond(episodeNotFoundView.contramap(lower))
+    )
+    <<< { userEpisodePermission >-> $0 } // Is there a name for this and/or do we want to introduce someting like "lift" to do it?
+    <<< filterMap(require3 >>> pure, or: loginAndRedirect)
+    <<< validateCreditRequest
+    <| applyCreditMiddleware
+
+private func applyCreditMiddleware<Z>(
+  _ conn: Conn<StatusLineOpen, T4<EpisodePermission, Episode, Database.User, Z>>
+  ) -> IO<Conn<ResponseEnded, Data>> {
+
+  let (episode, user) = (get2(conn.data), get3(conn.data))
+
+  guard user.episodeCreditCount > 0 else {
+    return conn
+      |> redirect(
+        to: .episode(.left(episode.slug)),
+        headersMiddleware: flash(.error, "You do not have any credits to use.")
+    )
+  }
+
+  return AppEnvironment.current.database.addEpisodeCredit(episode.sequence, user.id)
+    .flatMap { _ in
+      AppEnvironment.current.database.updateUser(user.id, nil, nil, nil, user.episodeCreditCount - 1)
+    }
+    .run
+    .flatMap(
+      either(
+        const(
+          conn
+            |> redirect(
+              to: .episode(.left(episode.slug)),
+              headersMiddleware: flash(.warning, "Something went wrong.")
+          )
+        ),
+        const(
+          conn
+            |> redirect(
+              to: .episode(.left(episode.slug)),
+              headersMiddleware: flash(.notice, "You now have access to this episode!")
+          )
+        )
+      )
+  )
+}
+
+private func validateCreditRequest<Z>(
+  _ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, T4<EpisodePermission, Episode, Database.User, Z>, Data>
+  ) -> Middleware<StatusLineOpen, ResponseEnded, T4<EpisodePermission, Episode, Database.User, Z>, Data> {
+
+  return { conn in
+    let (permission, episode, user) = (get1(conn.data), get2(conn.data), get3(conn.data))
+
+    guard user.episodeCreditCount > 0 else {
+      return conn
+        |> redirect(
+          to: .episode(.left(episode.slug)),
+          headersMiddleware: flash(.error, "You do not have any credits to use.")
+      )
+    }
+
+    guard isEpisodeViewable(for: permission) else {
+      return middleware(conn)
+    }
+
+    return conn
+      |> redirect(
+        to: .episode(.left(episode.slug)),
+        headersMiddleware: flash(.warning, "This episode is already available to you.")
+    )
+  }
+}
+
+private func userEpisodePermission<I, Z>(
+  _ conn: Conn<I, T4<Episode, Database.User?, Stripe.Subscription.Status?, Z>>
+  )
+  -> IO<Conn<I, T5<EpisodePermission, Episode, Database.User?, Stripe.Subscription.Status?, Z>>> {
+
+    let (episode, currentUser, subscriptionStatus) = (get1(conn.data), get2(conn.data), get3(conn.data))
```

Comment:
Don't we have a destructure function for this? At the very least with `_` for ignored args?

Reply:
the `lower` function only works on `TupleN` types, not `TN` types because it's not clear what should be done with the `Z` continuation piece. it's pretty annoying, but haven't figured out the best way to handle yet.